### DISCUSSION TITLE
How to handle anyio.WouldBlock in middleware?

### DISCUSSION THREAD

**Discussion by markedwards** [fastapi/fastapi#10387] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from starlette.middleware.base import BaseHTTPMiddleware, RequestResponseEndpoint
from starlette.requests import Request
from starlette.responses import Response


class DemoMiddleware(BaseHTTPMiddleware):

    async def dispatch(self, request: Request, call_next: RequestResponseEndpoint) -> Response:
        return await call_next(request)
```

### Description

I've seen `anyio.WouldBlock ` raised on some requests, with no pattern that is apparent to me.  It is followed by `anyio.EndOfStream` and finally `starlette.requests.ClientDisconnect`.

The trace of the first exception looks like:

```
ERROR:    Exception in ASGI application
Traceback (most recent call last):
  File "/venv/lib/python3.10/site-packages/anyio/streams/memory.py", line 98, in receive
    return self.receive_nowait()
  File "/venv/lib/python3.10/site-packages/anyio/streams/memory.py", line 93, in receive_nowait
    raise WouldBlock
anyio.WouldBlock
```

It's not clear to me what is causing a problem, or if there even is a problem.  And I have no clue what my middleware is supposed to do about it.

This is the only custom middleware, and is accompanied only by `fastapi.middleware.cors.CORSMiddleware`.

### Operating System

Linux

### Operating System Details

Running in a docker container based on `python:3.10-slim-bookworm`

### FastAPI Version

0.103.2

### Pydantic Version

2.4.2

### Python Version

3.10.13

**Comment by Kludex** [comment]
I think this will be solved when FastAPI bumps Starlette.

**Reply by ian-whitestone** [reply]
@Kludex wondering if there is a ballpark timeline for when Starlette version will get bumped?

I saw it's on the roadmap here: https://github.com/tiangolo/fastapi/issues/10370 but don't have a sense of timing from that.

Thanks!

**Reply by Kludex** [reply]
I don't know.

**Comment by markedwards** [comment]
Great, that's good to know.  Should I close this then, or do we need any additional confirmation?

**Reply by Kludex** [reply, marked as answer]
For housekeeping, and my organization, I would feel better to close, and check it again once Starlette gets bumped.

...but whatever you think is best is good for me.

**Comment by aryaniyaps** [comment]
any idea on when starlette will get bumped? I would really like to add my custom middleware without anyio errors!
Thank you!

**Reply by Kludex** [reply]
It was already.
